installinator/src/peers.rs

```diff
@@ -0,0 +1,334 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{fmt, future::Future, net::Ipv6Addr, str::FromStr, time::Duration};
+
+use anyhow::{bail, Result};
+use buf_list::BufList;
+use bytes::{Bytes, BytesMut};
+use display_error_chain::DisplayErrorChain;
+use itertools::Itertools;
+use progenitor_client::ResponseValue;
+use reqwest::StatusCode;
+use tokio::{
+    sync::{mpsc, oneshot},
+    time::Instant,
+};
+
+use crate::errors::{ArtifactFetchError, DiscoverPeersError};
+
+/// A fetched artifact
+pub(crate) struct FetchedArtifact {
+    pub(crate) attempt: usize,
+    pub(crate) addr: Ipv6Addr,
+    pub(crate) artifact: BufList,
+}
+
+impl FetchedArtifact {
+    /// In a loop, discover peers, and fetch from them.
+    ///
+    /// This only produces an error if the discover function errors out. In normal use, it is expected
+    /// to never error out.
```

Comment:
IMO this might be more clear as:

```suggestion
    /// If `discover_fn` returns an error other than `DiscoverPeersError::Retry`, that error is returned.
    /// Otherwise, this function continues looping until it succeeds.
```

Reply:
Thanks -- the comment was outdated; a previous version didn't have the retry/abort logic.